src/bsoncxx/string/view_or_value.hpp

```diff
@@ -0,0 +1,77 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <string>
+
+#include <bsoncxx/stdx/string_view.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <iostream>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace string {
+
+using string_variant = bsoncxx::view_or_value<stdx::string_view, std::string>;
+
+///
+/// Class representing a view-or-value variant type for strings.
+///
+class BSONCXX_API view_or_value : public string_variant {
+   public:
+    ///
+    /// Forward all constructors.
+    ///
+    using string_variant::string_variant;
+
+    ///
+    /// Construct a string::view_or_value using a null-terminated const char *.
+    ///
+    /// @param str A null-terminated string
+    ///
+    BSONCXX_INLINE view_or_value(const char* str)
+        : bsoncxx::view_or_value<stdx::string_view, std::string>(std::string(str)) {
```

Comment:
I got rid of the string_variant using statement, I didn't think we wanted it in a public header.  Should I add it back?

Re: whether `const char*` becomes a view or a value, I'm fine with it becoming, you're right that this makes more sense given how we handle `const std::string&`.  So the only way to construct an owning string::view_or_value is to std::move in a `std::string`.